### ISSUE TITLE
Upgrade to Next.js 15

### ISSUE THREAD

**Issue by wei** [wei/socialify#373]
https://nextjs.org/docs/app/building-your-application/upgrading/version-15

**Comment by shubhamkhan** [comment]
Hi @wei, /assign me this, I am interested to upgrade.

**Comment by KemingHe** [comment]
Hi, I'll be happy to deliver this in the next few days. Getting a branch going on my fork right now...

**Comment by wei** [comment]
@KemingHe We can migrate in one go, no need to separate.

**Comment by KemingHe** [comment]
> @KemingHe We can migrate in one go, no need to separate.

It's actually 3 separate migrations: 

- Next 14 to 15
- React 18 to 19
- Page router to app router

and only the last one is confirmed stable. Should I keep the page router when doing the first two steps (they are bundled together by NextJS)? What's your recommendation?

**Comment by wei** [comment]
> only the last one is confirmed stable

Can you explain what you mean? I believe [Next 15](https://nextjs.org/blog/next-15) and [React 19](https://react.dev/blog/2024/12/05/react-19) are both stable now. I'll let you decide the best workflow~

**Comment by KemingHe** [comment]
I see, okay. I will try a couple different ways and let you know then. 

<hr />

This is what I was referring to: 

Ref https://nextjs.org/docs/app/building-your-application/upgrading/version-15

```markdown
Good to know:

If you see a peer dependencies warning, you may need to update react and react-dom to the suggested versions, or you use the --force or --legacy-peer-deps flag to ignore the warning. This won't be necessary once both Next.js 15 and React 19 are stable.
```

and we are instructed to use  `react@rc react-dom@rc` (release candidate) instead of `react@latest`, meaning it's unstable.

**Comment by wei** [comment]
Sounds good, React 19 just turned stable so they probably need to update their docs.

I just noticed https://nextjs.org/docs/messages/no-stylesheets-in-head-component please also fix as part of the upgrade if you don't mind or I can do a separate PR as well.

**Comment by KemingHe** [comment]
> I just noticed https://nextjs.org/docs/messages/no-stylesheets-in-head-component please also fix as part of the upgrade if you don't mind or I can do a separate PR as well. 

No worries, this is already part of the `page router > app router` upgrade like you mentioned.

On it right now.

**Comment by KemingHe** [comment]
Hi @wei , I've been making great progress on the `page > app router` migration so far. All static elements (metadata, 404, landing page, header, footer...) plus unit tests are done.

 Would you like me to add API endpoint testing while I'm at it? (This will roughly double the delivery time but a must-have to ensure api consistency.)

Any recommendations you'd like me to be aware of?

**Comment by KemingHe** [comment]
Additional context: NextJS 14+ app router calls for complete API refactor, so that's why even if the api behaviors are fine (Playwright), I want a future-proof way of making sure they are internally consistent.

**Comment by wei** [comment]
That's fantastic! 

Let's add api testing in a separate pr. Can you create an issue with your proposal of how you plan to implement api testing?

**Comment by KemingHe** [comment]
Hello @wei , I think it'd be really helpful to the `page > app router` migration if we could talk through the architecture of Socialify over Zoom. What's your top 3 most available time slots in the next few days? I'm very flexible.

Hear from you soon!

**Comment by KemingHe** [comment]
Awesome news: API migration confirmed working, **AND** the simplest way to add testing is to use Playwright to expect match snapshot, since the endpoints mostly return images. 

Very close to full site migration to app router. Still would appreciate a Zoom meeting to speed things up.

**Comment by wei** [comment]
Excellent, let's coordinate over dm

**Comment by KemingHe** [comment]
Talk to you soon in LinkedIn. 🌻

**Comment by wei** [comment]
Resolved in #450 🚀
